File: USACO/Contests/1819_1G/Dining.java
import java.io.*;
import java.util.*;

public class Dining {
	static final boolean stdin = false;
	static final String filename = "dining";
	static Reader in;
	static PrintWriter out;
	
	static class Solver {
		static final int BIG=(int)1e9;
		int n;
		List<PII>[] adj;
		int[] y, d1;
		PII[] d2;
		
		Solver() {
			//input
			n=in.nextInt();
			int m=in.nextInt(), k=in.nextInt();
			adj = new List[n];
			for(int i=0; i<n; ++i)
				adj[i] = new ArrayList<PII>();
			while(m-->0) {
				int a=in.nextInt()-1, b=in.nextInt()-1, t=in.nextInt();
				adj[a].add(new PII(t, b));
				adj[b].add(new PII(t, a));
			}
			y = new int[n];
			while(k-->0) {
				int a=in.nextInt()-1, b=in.nextInt();
				y[a]=Math.max(b, y[a]);
			}
			dijkstra1();
			dijkstra2();
			//output
			//if there is a haybale at node n, all answers are 1
			for(int i=0; i<n-1; ++i)
				out.println((d2[i].b<n-1||y[n-1]>0)?1:0);
		}
		
		void dijkstra1() {
			//first dijkstra from n
			PriorityQueue<PII> pq = new PriorityQueue<PII>();
			d1 = new int[n];
			Arrays.fill(d1, Integer.MAX_VALUE);
			d1[n-1]=0;
			pq.add(new PII(0, n-1));
			while(!pq.isEmpty()) {
				PII u=pq.poll();
				if(u.a>d1[u.b])
					continue;
				for(PII v : adj[u.b]) {
					if(d1[v.b]>u.a+v.a) {
						d1[v.b]=u.a+v.a;
						pq.add(new PII(d1[v.b], v.b));
					}
				}
			}
		}
		
		void dijkstra2() {
			//multi-source dijkstra
			//we also keep track of the source
			//sources that are not node n mean that haybales are possible
			//distances are stored as a pair {distance, source}
			PriorityQueue<PPIII> pq = new PriorityQueue<PPIII>();
			d2 = new PII[n];
			Arrays.fill(d2, new PII(Integer.MAX_VALUE, n));
			d2[n-1]=new PII(BIG, n-1);
			pq.add(new PPIII(d2[n-1], n-1));
			//add other nodes with haybales as sources
			for(int i=0; i<n-1; ++i) {
				if(y[i]<=0)
					continue;
				d2[i]=new PII(d1[i]+BIG-y[i], i);
				pq.add(new PPIII(d2[i], i));
			}
			while(!pq.isEmpty()) {
				PPIII u=pq.poll();
				if(u.a.compareTo(d2[u.b])>0)
					continue;
				for(PII v : adj[u.b]) {
					if(d2[v.b].compareTo(new PII(u.a.a+v.a, u.a.b))>0) {
						d2[v.b]=new PII(u.a.a+v.a, u.a.b);
						pq.add(new PPIII(d2[v.b], v.b));
					}
				}
			}
		}
		
		static class PII implements Comparable<PII> {
			int a, b;
			PII(int a, int b) {
				this.a=a;
				this.b=b;
			}
			//Lexicographical comparison
			public int compareTo(PII o) {
				return a==o.a?b-o.b:a-o.a;
			}
		}
		
		static class PPIII implements Comparable<PPIII> {
			PII a;
			int b;
			PPIII(PII a, int b) {
				this.a=a;
				this.b=b;
			}
			//Lexicographical comparison
			public int compareTo(PPIII o) {
				return a.equals(o.a)?b-o.b:a.compareTo(o.a);
			}
		}
	}
	
	
	public static void main(String[] args) throws Exception {
		if(stdin) {
			in = new Reader();
			out = new PrintWriter(System.out);
		} else {
			in = new Reader(filename+".in");
			out = new PrintWriter(filename+".out");
		}
		new Solver();
		out.close();
	}
	
	static class Reader {
		BufferedReader br;
		StringTokenizer st;
		Reader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}
		Reader(String filename) throws Exception {
			br = new BufferedReader(new FileReader(filename));
		}
		String next() {
			while(st==null||!st.hasMoreTokens()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch(Exception e) {}
			}
			return st.nextToken();
		}
		int nextInt() {
			return Integer.parseInt(next());
		}
	}
}

File: USACO/Contests/1819_1G/Dining.cpp
#include <bits/stdc++.h>
using namespace std;

#define a array

const int mxN=5e4, BIG=1e9;
int n, m, k, y[mxN], ai, bi, ti, d1[mxN];
a<int, 2> d2[mxN];
vector<a<int, 2>> adj[mxN];

void dijkstra1() {
	//first dijkstra from n
	priority_queue<a<int, 2>, vector<a<int, 2>>, greater<a<int, 2>>> pq;
	memset(d1, 0x7f, 4*n);
	d1[n-1]=0;
	pq.push({0, n-1});
	while(pq.size()) {
		a<int, 2> u=pq.top();
		pq.pop();
		if(u[0]>d1[u[1]])
			continue;
		for(a<int, 2> v : adj[u[1]]) {
			if(d1[v[1]]>u[0]+v[0]) {
				d1[v[1]]=u[0]+v[0];
				pq.push({d1[v[1]], v[1]});
			}
		}
	}
}

void dijkstra2() {
	//multi-source dijkstra
	//we also keep track of the source
	//sources that are not node n mean that haybales are possible
	//distances are stored as a pair {distance, source}
	priority_queue<a<int, 3>, vector<a<int, 3>>, greater<a<int, 3>>> pq;
	memset(d2, 0x7f, sizeof(d2[0])*n);
	d2[n-1]={BIG, n-1};
	pq.push({BIG, n-1, n-1});
	//add other nodes with haybales as sources
	for(int i=0; i<n-1; ++i) {
		if(!y[i])
			continue;
		d2[i]={d1[i]+BIG-y[i], i};
		pq.push({d1[i]+BIG-y[i], i, i});
	}
	while(pq.size()) {
		a<int, 3> u=pq.top();
		pq.pop();
		if(a<int, 2>{u[0], u[1]}>d2[u[2]])
			continue;
		for(a<int, 2> v : adj[u[2]]) {
			if(d2[v[1]]>a<int, 2>{u[0]+v[0], u[1]}) {
				d2[v[1]]={u[0]+v[0], u[1]};
				pq.push({u[0]+v[0], u[1], v[1]});
			}
		}
	}
}

int main() {
	ifstream cin("dining.in");
	ofstream cout("dining.out");
	
	//input
	cin >> n >> m >> k;
	while(m--) {
		cin >> ai >> bi >> ti, --ai, --bi;
		adj[ai].push_back({ti, bi});
		adj[bi].push_back({ti, ai});
	}
	while(k--) {
		cin >> ai >> bi, --ai;
		y[ai]=max(bi, y[ai]);
	}
	dijkstra1();
	dijkstra2();
	//output
	//if there is a haybale at node n, all answers are 1
	for(int i=0; i<n-1; ++i)
		cout << (d2[i][1]<n-1||y[n-1]) << "\n";
}
